app/constants/mods.py

```diff
@@ -53,113 +54,96 @@ def __repr__(self) -> str:
         if self.value == Mods.NOMOD:
             return "NM"
 
-        mod_str = []
-        _dict = mod2modstr_dict  # global
-
-        for mod in Mods:
-            if self.value & mod:
-                mod_str.append(_dict[mod])
-
+        mod_str = [mod2modstr_dict[mod] for mod in Mods if self.value & mod]
         return "".join(mod_str)
 
     def filter_invalid_combos(self, mode_vn: int) -> Mods:
         """Remove any invalid mod combinations."""
-
-        # 1. mode-inspecific mod conflictions
-        _dtnc = self & (Mods.DOUBLETIME | Mods.NIGHTCORE)
-        if _dtnc == (Mods.DOUBLETIME | Mods.NIGHTCORE):
-            self &= ~Mods.DOUBLETIME  # DTNC
-        elif _dtnc and self & Mods.HALFTIME:
-            self &= ~Mods.HALFTIME  # (DT|NC)HT
-
-        if self & Mods.EASY and self & Mods.HARDROCK:
-            self &= ~Mods.HARDROCK  # EZHR
-
-        if self & (Mods.NOFAIL | Mods.RELAX | Mods.AUTOPILOT):
-            if self & Mods.SUDDENDEATH:
-                self &= ~Mods.SUDDENDEATH  # (NF|RX|AP)SD
-            if self & Mods.PERFECT:
-                self &= ~Mods.PERFECT  # (NF|RX|AP)PF
-
-        if self & (Mods.RELAX | Mods.AUTOPILOT):
-            if self & Mods.NOFAIL:
-                self &= ~Mods.NOFAIL  # (RX|AP)NF
-
-        if self & Mods.PERFECT and self & Mods.SUDDENDEATH:
-            self &= ~Mods.SUDDENDEATH  # PFSD
-
-        # 2. remove mode-unique mods from incorrect gamemodes
-        if mode_vn != 0:  # osu! specific
+        # define the mapping between conflicting mod combinations and their resolution
+        conflicts = {
+            (Mods.DOUBLETIME, Mods.NIGHTCORE): Mods.DOUBLETIME,  # DTNC
+            (
+                Mods.DOUBLETIME,
+                Mods.NIGHTCORE,
+                Mods.HALFTIME,
+            ): Mods.NIGHTCORE,  # (DT|NC)HT
+            (Mods.EASY, Mods.HARDROCK): Mods.EASY,  # EZHR
+            (
+                Mods.NOFAIL,
+                Mods.RELAX,
+                Mods.AUTOPILOT,
+                Mods.SUDDENDEATH,
+            ): Mods.NOFAIL,  # (NF|RX|AP)SD
+            (
+                Mods.NOFAIL,
+                Mods.RELAX,
+                Mods.AUTOPILOT,
+                Mods.PERFECT,
+            ): Mods.NOFAIL,  # (NF|RX|AP)PF
+            (Mods.RELAX, Mods.AUTOPILOT, Mods.NOFAIL): Mods.RELAX,  # (RX|AP)NF
+            (Mods.PERFECT, Mods.SUDDENDEATH): Mods.PERFECT,  # PFSD
+            (Mods.HIDDEN, Mods.FADEIN): Mods.HIDDEN,  # HDFI
+        }
+
+        # check for conflicting mod combinations and apply the resolution
+        for conflict, resolution in conflicts.items():
```

Comment:
the name "resolution" to me sounded more like the end result, but it's actually what's removed from the end result

seems a bit counterintuitively named

Reply:
it makes sense to me because the removal of the invalid combination is the resolution, but i can get why someone could get confused about it

i got no idea how to name it though